racket/src/thread/thread.rkt

```diff
@@ -556,13 +557,16 @@
   ;; awoken or there was already mail.
   (define t (current-thread))
   (let loop ()
-    (cond
-      [(is-mail? t)
-       (atomically
-        (dequeue-mail! t))]
-      [else
-       ((thread-internal-suspend! t #f void block))
-       (loop)])))
+    ((atomically
+      (cond
+        [(is-mail? t)
+         (let ([msg (dequeue-mail! t)])
+           (lambda () msg))]
+        [else
+         (let ([f (thread-internal-suspend! t #f void block)])
+           (lambda ()
+             (f)
+             (loop)))])))))
```

Comment:
Removing the first block of code will get rid of this line 
```
(set-thread-waiting-mail?! t #t)
``` 
which will cause the code to stop working.    You could move that line of code into the let loop.